File: server/app/services/chat/context/assembly.py
"""
Context assembly module for building LLM input messages.

This module handles the assembly of context components into a unified message
format for LLM processing. It implements the "one big message" pattern where
character settings, background story, and recent messages are combined into
a single prompt.

The assembly process includes:
- Character settings (personality, style, identity) integration
- Background story formatting with metadata
- Recent message formatting with sequence numbers
- User state injection as natural language in instruction area
- Agent delivery integration for world-interaction results

Template Design:
- Uses narrative-style section headers instead of bracketed labels
- "Background context from earlier" and "Recent conversation" create temporal flow
- Metadata (message numbers) preserved for debugging and context clarity
- User state placed in instruction area (after narrative, before response directive)
  to preserve narrative flow while guiding response strategy
"""

from typing import List, Dict, Any, Optional
from langchain_core.messages import HumanMessage, AIMessage, BaseMessage
from app.services.dto.task_result import TaskResult
from app.logger import logger


class ContextAssemblyService:
    """
    Service for assembling context components into LLM-ready messages.
    
    This service implements the context assembly strategy where:
    - Character settings define the agent's personality and style
    - Summary and recent messages are decoupled (may overlap)
    - Metadata labels help LLM understand the scope of each component
    - Background story provides compressed historical context
    - Recent messages provide precise details for current context
    - User state guides response strategy via natural language instruction
    - Agent delivery provides world-interaction results when applicable
    """
    
    # Template for full context with background story and character settings
    ONE_BIG_MESSAGE_TEMPLATE = """{character_section}Background context from earlier in the conversation (messages 1-{summary_version}):

{background_story}

---

Recent conversation (messages {recent_start}-{recent_end}):

{dialog_section}

---

{task_result_section}{user_state_instruction}Please respond directly to the user. Do not use parenthetical action descriptions or non-verbal content."""

    # Template for simple context without background story (V < N case)
    ONE_BIG_MESSAGE_NO_STORY_TEMPLATE = """{character_section}Conversation record (messages {recent_start}-{recent_end}):

{dialog_section}

---

{task_result_section}{user_state_instruction}Please respond directly to the user. Do not use parenthetical action descriptions or non-verbal content."""

    @staticmethod
    def _format_character_section(character_settings: Optional[str]) -> str:
        """
        Format character settings section for the prompt.
        
        Args:
            character_settings: Agent's personality, style, and identity settings
            
        Returns:
            Formatted character section string
        """
        if not character_settings:
            return ""
        return f"[Your Character]\n{character_settings}\n\n---\n\n"

    @staticmethod
    def _format_user_state_instruction(user_state_description: Optional[str]) -> str:
        """
        Format user state as natural language instruction.
        
        User state is placed in the instruction area (after narrative sections,
        before response directive) to preserve narrative flow while guiding
        the LLM's response strategy.
        
        Args:
            user_state_description: Natural language description of user state,
                e.g. "The user appears frustrated, is seeking help with a problem,
                and is likely at work on desktop."
            
        Returns:
            Formatted instruction string, or empty string if no user state
        """
        if not user_state_description:
            return ""
        return f"{user_state_description}\nAdjust your response tone, detail level, and strategy accordingly.\n\n"

    @staticmethod
    def _format_task_result_section(task_result: Optional["TaskResult"]) -> str:
        """
        Format agent delivery section for the prompt.
        
        When the agent has executed a world-interaction task, this section
        provides the execution results to the LLM for generating a natural
        language response.
        
        Args:
            task_result: The agent execution result containing status,
                result/reason, and notes.
            
        Returns:
            Formatted agent delivery section string, or empty string if no delivery.
        """
        if not task_result:
            return ""
        
        if task_result.status == "success":
            return f"[Task Execution Result]\nThe following task was completed successfully:\n\n{task_result.result or 'Task completed.'}\n\n---\n\n"
        else:
            # Failure case: include notes for context
            notes_section = ""
            if task_result.notes:
                notes_section = f"\n\nProgress notes:\n{task_result.notes}"
            
            return f"[Task Execution Interrupted]\nThe task could not be completed.\n\nReason: {task_result.reason or 'Unknown error'}{notes_section}\n\n---\n\n"

    @staticmethod
    def assemble_context(
        character_settings: Optional[str],
        background_story: Optional[str],
        recent_messages: List[Dict[str, Any]],
        summary_version: Optional[int] = None,
        recent_start: int = 1,
        recent_end: int = 1,
        user_state_description: Optional[str] = None,
        task_result: Optional["TaskResult"] = None,
    ) -> List[BaseMessage]:
        """
        Assemble context into one big message for LLM processing.
        
        This method combines character settings, background story, and recent messages
        into a unified message format. The background story and recent messages
        are decoupled - they may overlap in coverage.
        
        Args:
            character_settings: Agent's personality, style, and identity settings
            background_story: Background narrative from summary + RAG segments
            recent_messages: Recent completed messages (including trigger_message as the latest)
            summary_version: Version number of the summary (covers messages 1 to summary_version)
            recent_start: Starting message sequence number for recent messages
            recent_end: Ending message sequence number for recent messages
            user_state_description: Natural language description of inferred user state,
                placed in instruction area to guide response strategy
            task_result: Agent execution result for world-interaction tasks,
                provides execution status and results for LLM to formulate response
            
        Returns:
            List of LangChain messages ready for LLM processing
        """
        messages = []

        # Format character settings section
        character_section = ContextAssemblyService._format_character_section(character_settings)

        # Format user state instruction
        user_state_instruction = ContextAssemblyService._format_user_state_instruction(user_state_description)

        # Format agent delivery section
        task_result_section = ContextAssemblyService._format_task_result_section(task_result)

        # Format recent messages into dialog section
        dialog_lines = []
        for msg in recent_messages:
            role = "User" if msg["role"] == "user" else "You"
            dialog_lines.append(f"{role}: {msg['content']}")
        dialog_section = "\n".join(dialog_lines)

        # Choose template based on whether background story exists
        if background_story and summary_version:
            one_big_message = ContextAssemblyService.ONE_BIG_MESSAGE_TEMPLATE.format(
                character_section=character_section,
                background_story=background_story,
                dialog_section=dialog_section,
                summary_version=summary_version,
                recent_start=recent_start,
                recent_end=recent_end,
                user_state_instruction=user_state_instruction,
                task_result_section=task_result_section,
            )
        else:
            one_big_message = ContextAssemblyService.ONE_BIG_MESSAGE_NO_STORY_TEMPLATE.format(
                character_section=character_section,
                dialog_section=dialog_section,
                recent_start=recent_start,
                recent_end=recent_end,
                user_state_instruction=user_state_instruction,
                task_result_section=task_result_section,
            )

        # Add the one big message as a HumanMessage
        messages.append(HumanMessage(content=one_big_message))

        logger.info(f"Assembled context with {len(messages)} messages, user_state: {user_state_description is not None}, task_result: {task_result is not None}")
        return messages

File: server/internal/service/chat/chatctx/assembly.go
// Package context implements the context engineering pipeline for chat processing.
//
// This package provides the context assembly services that build the LLM message
// sequence from various context sources: summaries, narratives, retrieval results,
// user state, and task results. It matches Python's chat/context module.
package chatctx

import (
	"fmt"
	"strings"

	"private-buddy-server/internal/model"
	"private-buddy-server/internal/service/llm"

	applogger "private-buddy-server/internal/logger"
)

// Template for full context with background story, segments, and character settings.
// Uses narrative-style section headers instead of bracketed labels:
//   - "Background context from earlier" and "Recent conversation" create temporal flow
//   - Segments section uses narrative transition to reduce abruptness
//   - Metadata (message numbers) preserved for debugging and context clarity
//   - User state placed in instruction area (after narrative, before response directive)
//     to preserve narrative flow while guiding response strategy
const oneBigMessageTemplate = `%sBackground context from earlier in the conversation (messages 1-%d):

%s

%s---

Recent conversation (messages %d-%d):

%s

---

%s%sPlease respond directly to the user. Do not use parenthetical action descriptions or non-verbal content.`

// Template for simple context without background story (V < N case).
// Used when there are not enough messages to generate a summary.
// Segments section is included when KB-retrieved content is available.
const oneBigMessageNoStoryTemplate = `%sConversation record (messages %d-%d):

%s

---

%s%s%sPlease respond directly to the user. Do not use parenthetical action descriptions or non-verbal content.`

// TaskResultForAssembly represents the task execution result for context assembly.
// Mirrors Python's TaskResult DTO used in context assembly.
// Status is "success" or "failure"; Result/Reason/Notes are populated accordingly.
type TaskResultForAssembly struct {
	Status string `json:"status"`
	Result string `json:"result"`
	Reason string `json:"reason"`
	Notes  string `json:"notes"`
}

// formatCharacterSection formats character settings section for the prompt.
// Returns "[Your Character]\n{settings}\n\n---\n\n" or empty string if nil/empty.
func formatCharacterSection(characterSettings string) string {
	if characterSettings == "" {
		return ""
	}
	return fmt.Sprintf("[Your Character]\n%s\n\n---\n\n", characterSettings)
}

// formatSegmentsSection formats relevant segments as an independent section.
// Segments are RAG-retrieved historical fragments placed with narrative transition,
// since they could not be fused into the pre-generated cached narrative.
// Returns "Some additional details from earlier conversations...\n{items}\n\n" or empty string.
func formatSegmentsSection(relevantSegments []Segment) string {
	if len(relevantSegments) == 0 {
		return ""
	}

	var segmentsText []string
	for _, seg := range relevantSegments {
		sourceLabel := "ChatHistory"
		if seg.Source == SourceKnowledgeBase {
			sourceLabel = "KnowledgeBase"
		}
		segmentsText = append(segmentsText, fmt.Sprintf("- (%s) %s", sourceLabel, seg.Content))
	}

	return fmt.Sprintf("Some additional details from earlier conversations that may be relevant:\n\n%s\n\n", strings.Join(segmentsText, "\n"))
}

// formatUserStateInstruction formats user state as natural language instruction.
// Placed in the instruction area (after narrative, before response directive)
// to preserve narrative flow while guiding response strategy.
// Returns "{description}\nAdjust your response tone, detail level, and strategy accordingly.\n\n" or empty string.
func formatUserStateInstruction(userStateDescription string) string {
	if userStateDescription == "" {
		return ""
	}
	return fmt.Sprintf("%s\nAdjust your response tone, detail level, and strategy accordingly.\n\n", userStateDescription)
}

// formatTaskResultSection formats agent delivery section for the prompt.
// Provides execution status and results for LLM to formulate response:
//   - success: "[Task Execution Result]\nThe following task was completed successfully:\n\n{result}\n\n---\n\n"
//   - failure: "[Task Execution Interrupted]\nThe task could not be completed.\n\nReason: {reason}\n\n---\n\n"
func formatTaskResultSection(taskResult *TaskResultForAssembly) string {
	if taskResult == nil {
		return ""
	}

	if taskResult.Status == "success" {
		result := "Task completed."
		if taskResult.Result != "" {
			result = taskResult.Result
		}
		return fmt.Sprintf("[Task Execution Result]\nThe following task was completed successfully:\n\n%s\n\n---\n\n", result)
	}

	notesSection := ""
	if taskResult.Notes != "" {
		notesSection = fmt.Sprintf("\n\nProgress notes:\n%s", taskResult.Notes)
	}

	reason := "Unknown error"
	if taskResult.Reason != "" {
		reason = taskResult.Reason
	}

	return fmt.Sprintf("[Task Execution Interrupted]\nThe task could not be completed.\n\nReason: %s%s\n\n---\n\n", reason, notesSection)
}

// AssembleContext assembles context into one big message for LLM processing.
//
// This method combines character settings, background story (cached narrative),
// relevant segments, and recent messages into a unified message format.
//
// The background story is a cached narrative generated in background alongside
// the summary. Segments are RAG-retrieved fragments placed as an independent
// section with narrative transition, since they could not be fused into the
// pre-generated narrative.
//
// Parameters:
//   - characterSettings: agent's personality, style, and identity settings
//   - backgroundStory: cached narrative from summary record
//   - recentMessages: recent completed messages (including trigger_message as the latest)
//   - relevantSegments: RAG-retrieved historical segments (independent section)
//   - summaryVersion: version number of the summary (covers messages 1 to summaryVersion)
//   - recentStart: starting message sequence number for recent messages
//   - recentEnd: ending message sequence number for recent messages
//   - userStateDescription: natural language description of inferred user state,
//     placed in instruction area to guide response strategy
//   - taskResult: agent execution result for world-interaction tasks,
//     provides execution status and results for LLM to formulate response
func AssembleContext(
	characterSettings string,
	backgroundStory string,
	recentMessages []model.Message,
	relevantSegments []Segment,
	summaryVersion int,
	recentStart int,
	recentEnd int,
	userStateDescription string,
	taskResult *TaskResultForAssembly,
) []llm.Message {
	characterSection := formatCharacterSection(characterSettings)
	userStateInstruction := formatUserStateInstruction(userStateDescription)
	taskResultSection := formatTaskResultSection(taskResult)

	var dialogLines []string
	for _, msg := range recentMessages {
		role := "User"
		if msg.Role != "user" {
			role = "You"
		}
		dialogLines = append(dialogLines, fmt.Sprintf("%s: %s", role, msg.Content))
	}
	dialogSection := strings.Join(dialogLines, "\n")

	var oneBigMessage string
	segmentsSection := formatSegmentsSection(relevantSegments)
	if backgroundStory != "" && summaryVersion != -1 {
		oneBigMessage = fmt.Sprintf(oneBigMessageTemplate,
			characterSection,
			summaryVersion,
			backgroundStory,
			segmentsSection,
			recentStart,
			recentEnd,
			dialogSection,
			taskResultSection,
			userStateInstruction,
		)
	} else {
		oneBigMessage = fmt.Sprintf(oneBigMessageNoStoryTemplate,
			characterSection,
			recentStart,
			recentEnd,
			dialogSection,
			segmentsSection,
			taskResultSection,
			userStateInstruction,
		)
	}

	messages := []llm.Message{
		{Role: "user", Content: oneBigMessage},
	}

	applogger.L.Info("Assembled context",
		"message_count", len(messages),
		"has_user_state", userStateDescription != "",
		"has_task_result", taskResult != nil,
		"segments", len(relevantSegments),
	)

	return messages
}
